### PR TITLE
Issue #543 write behind config defaults and legal values

### PR DESCRIPTION
**Pull request by ljacomet**
First part for issue #543, addresses defaults and legal values for:
* queue size
* rate limit
* min and max write delay

Follow ups will be added here or in a new PR if this gets merged before.

### CODE REVIEW

File: core/src/test/java/org/ehcache/config/writebehind/DefaultWriteBehindConfigurationTest.java [hunk 1]
```diff
@@ -18,6 +18,9 @@
 
 import org.junit.Test;
 
+import java.math.BigInteger;
```
Comment (albinsuresh) [inline core/src/test/java/org/ehcache/config/writebehind/DefaultWriteBehindConfigurationTest.java]: Unused

File: impl/src/test/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindTestBase.java [hunk 2]
```diff
@@ -66,10 +66,13 @@ public void testBulkWrites() throws Exception {
     
     Map<String, String> entries = new HashMap<String, String>(10);
     Set<String> keys = new HashSet<String>(10);
-    for(int i=10 ; i<20; i++) {
+    for(int i=10 ; i<19; i++) {
       entries.put("test"+i, "test"+i);
       keys.add("test"+i);
     }
+    Thread.sleep(1000);
```
Comment (albinsuresh) [inline impl/src/test/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindTestBase.java]: Thread.sleep()

File: impl/src/test/java/org/ehcache/loaderwriter/writebehind/WriteBehindTestLoaderWriter.java [hunk 3]
```diff
@@ -83,7 +83,7 @@ public void write(K key, V value) throws Exception {
 
   @Override
   public void writeAll(Iterable<? extends Entry<? extends K, ? extends V>> entries) throws BulkCacheWritingException, Exception {
-
+    System.out.println("WriteAll " + entries);
```
Comment (albinsuresh) [inline impl/src/test/java/org/ehcache/loaderwriter/writebehind/WriteBehindTestLoaderWriter.java]: sysout

File: impl/src/test/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindTestBase.java (hunk 2 above)
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindTestBase.java]: Should not push PRs late at night - that needs removed

File: impl/src/test/java/org/ehcache/loaderwriter/writebehind/WriteBehindTestLoaderWriter.java (hunk 3 above)
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/loaderwriter/writebehind/WriteBehindTestLoaderWriter.java]: Same here

**Comment by ljacomet** [comment]
test this please

**Comment by ljacomet** [comment]
Updated the PR with changes related to
* batch size
* retry attempts

File: core/src/main/java/org/ehcache/config/writebehind/DefaultWriteBehindConfiguration.java [hunk 4]
```diff
@@ -28,15 +28,14 @@
...
-  private int maxWriteDelay = 1;
-  private int rateLimitPerSecond = 0;
+  private int maxWriteDelay = Integer.MAX_VALUE;
+  private int rateLimitPerSecond = Integer.MAX_VALUE;
   private boolean writeCoalescing = false;
-  private boolean writeBatching = false;
   private int writeBatchSize = 1;
-  private int retryAttempts = 1;
+  private int retryAttempts = 0;
   private int retryAttemptDelaySeconds = 1;
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/config/writebehind/DefaultWriteBehindConfiguration.java]: Any suggestion for having the default here different than 1?

Comment (chrisdennis) [inline core/src/main/java/org/ehcache/config/writebehind/DefaultWriteBehindConfiguration.java]: No default, and no error in the place of default zero retry attempts.  If you specify a non-zero retry attempts value then you must also specify the delay you want: time to break out the pink library.
Schematron!

File: core/src/main/java/org/ehcache/config/writebehind/DefaultWriteBehindConfiguration.java [hunk 5]
```diff
@@ -91,26 +85,27 @@ public int getWriteBehindMaxQueueSize() {
...
-    }
-    if (maxWriteDelay < 1) {
-      throw new IllegalArgumentException("Maximum write delay seconds cannot be less than 1.");
-    }
-    if (maxWriteDelay < minWriteDelay) {
-      throw new IllegalArgumentException(
-          format("Maximum write delay (%d seconds) must be equal or greater than minimum write delay (%d seconds)", maxWriteDelay, minWriteDelay));
+      throw new IllegalArgumentException("Minimum write delay seconds cannot be less than 0");
+    } else if (minWriteDelay > maxWriteDelay) {
+      throw new IllegalArgumentException("Minimum write delay must be smaller than maximum write delay");
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/config/writebehind/DefaultWriteBehindConfiguration.java]: smaller than or equal to

File: core/src/main/java/org/ehcache/config/writebehind/DefaultWriteBehindConfiguration.java [hunk 6]
```diff
@@ -91,26 +85,27 @@ public int getWriteBehindMaxQueueSize() {
...
     }
     this.minWriteDelay = minWriteDelay;
+  }
+
+  public void setMaxWriteDelay(int maxWriteDelay) {
+    if (maxWriteDelay < 0) {
+      throw new IllegalArgumentException("Maximum write delay cannot be less than 1");
+    } else if (maxWriteDelay < minWriteDelay) {
+      throw new IllegalArgumentException("Maximum write delay must be larger than minimum write delay");
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/config/writebehind/DefaultWriteBehindConfiguration.java]: larger than or equal to

File: impl/src/test/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindTestBase.java [hunk 7]
```diff
@@ -70,7 +70,7 @@ public void testBulkWrites() throws Exception {
       entries.put("test"+i, "test"+i);
       keys.add("test"+i);
     }
-    
+
```
Comment (chrisdennis) [inline impl/src/test/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindTestBase.java]: huh?

File: impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java [hunk 8]
```diff
@@ -359,14 +355,12 @@ private void processItems() throws RuntimeException {
...
-              waitUntilEnoughTimeHasPassed(quarantinedItems, batchSize, secondsSinceLastWorkDone);
-              return;
-            }
+          final long secondsSinceLastWorkDone = (System.currentTimeMillis() - lastWorkDone.get()) / MS_IN_SEC;
+          final long maxBatchSizeSinceLastWorkDone = rateLimitPerSecond * secondsSinceLastWorkDone;
+          final int batchSize = determineBatchSize(quarantinedItems);
+          if (batchSize > maxBatchSizeSinceLastWorkDone) {
+            waitUntilEnoughTimeHasPassed(quarantinedItems, batchSize, secondsSinceLastWorkDone);
+            return;
           }
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java]: No rate limiting when batch size is 1?

File: xml/src/main/resources/ehcache-core.xsd [hunk 9]
```diff
@@ -213,34 +213,17 @@
...
-        </xs:sequence>
-        <xs:attribute type="xs:positiveInteger" name="concurrency" default="1" use="optional"/>
-        <xs:attribute type="xs:nonNegativeInteger" name="size" default="0" use="optional"/>
-        <xs:attribute type="xs:boolean" name="coalesce" use="optional"/>
-      </xs:complexType>
-    </xs:element>
+          <xs:attribute type="xs:positiveInteger" name="concurrency" default="1" use="optional"/>
+          <xs:attribute type="xs:positiveInteger" name="size" default="2147483647" use="optional"/>
+          <xs:attribute type="xs:boolean" name="coalesce" use="optional"/>
+          <xs:attribute name="ratelimitpersecond" type="xs:positiveInteger" default="2147483647" use="optional"/>
```
Comment (chrisdennis) [inline xml/src/main/resources/ehcache-core.xsd]: Camel Case?

File: impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java [hunk 10]
```diff
@@ -392,7 +386,7 @@ private void processItems() throws RuntimeException {
   private void processQuarantinedItems(List<SingleOperation<K, V>> quarantinedItems) throws Exception {
     LOGGER.debug("{} : processItems() : processing " + " quarantined items", getThreadName());
 
-    if (writeBatching && writeBatchSize > 0) {
+    if (writeBatchSize > 1) {
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java]: I think this if block could be pushed deeper here, and then we can extract out the commonality between `processBatchedOperations(...)` and `processSingleOperation(...)`

File: core/src/main/java/org/ehcache/config/writebehind/DefaultWriteBehindConfiguration.java (hunk 4 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/config/writebehind/DefaultWriteBehindConfiguration.java]: Decided that retry will remain an element with two mandatory attributes

File: xml/src/main/resources/ehcache-core.xsd [hunk 11]
```diff
@@ -214,33 +214,22 @@
...
-              </xs:simpleContent>
-            </xs:complexType>
-          </xs:element>
-        </xs:sequence>
-        <xs:attribute type="xs:positiveInteger" name="concurrency" default="1" use="optional"/>
-        <xs:attribute type="xs:nonNegativeInteger" name="size" default="0" use="optional"/>
-        <xs:attribute type="xs:boolean" name="coalesce" use="optional"/>
-      </xs:complexType>
-    </xs:element>
+            <xs:element name="retry" minOccurs="0" maxOccurs="1">
```
Comment (ljacomet) [inline xml/src/main/resources/ehcache-core.xsd]: Kept retry as an element to indicate that both attributes must be provided together.

File: impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java (hunk 8 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java]: Created #547 to address this.

File: impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java (hunk 10 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/loaderwriter/writebehind/AbstractWriteBehindQueue.java]: Created #547 to address this.

File: impl/src/test/java/org/ehcache/loaderwriter/writebehind/WriteBehindTestLoaderWriter.java [hunk 12]
```diff
@@ -83,7 +83,6 @@ public void write(K key, V value) throws Exception {
 
   @Override
   public void writeAll(Iterable<? extends Entry<? extends K, ? extends V>> entries) throws BulkCacheWritingException, Exception {
-
```
Comment (albinsuresh) [inline impl/src/test/java/org/ehcache/loaderwriter/writebehind/WriteBehindTestLoaderWriter.java]: I don't wanna be a pedant here ;)

Comment (ljacomet) [inline impl/src/test/java/org/ehcache/loaderwriter/writebehind/WriteBehindTestLoaderWriter.java]: Fixed!

**Comment by ljacomet** [comment]
test this please

**Pull request merged**